### PR TITLE
srds: remove to-de-removed scopes first and then apply additions to avoid scope key conflict.

### PR DESCRIPTION
**Pull request by stevenzzzz**
Description: remove scopes first then add scopes to avoid scope key conflict when a new scope has the same key as an existing scope

Risk Level: MID
Testing: unit test
Docs Changes:
Release Notes:
Fixes: #9349

### CODE REVIEW

File: test/common/router/scoped_rds_test.cc [hunk 1]
```diff
@@ -477,6 +477,70 @@ route_configuration_name: foo_routes
             "foo_routes");
 }
 
+// Tests that scope-key conflict resources in different pushes are handled correctly.
+TEST_F(ScopedRdsTest, ScopeKeyReuseInDifferentPushes) {
```
Comment (AndresGuedez) [inline test/common/router/scoped_rds_test.cc]: nit:s/pushes/config updates

File: test/common/router/scoped_rds_test.cc [hunk 2]
```diff
@@ -477,6 +477,70 @@ route_configuration_name: foo_routes
...
+  EXPECT_EQ(getScopedRouteMap().size(), 1UL);
+  EXPECT_EQ(getScopedRouteMap().count("foo_scope"), 0);
+  EXPECT_EQ(getScopedRouteMap().count("foo_scope2"), 1);
+  // The same scope-key now points to the same route table.
+  EXPECT_EQ(getScopedRdsProvider()
+                ->config<ScopedConfigImpl>()
+                ->getRouteConfig(TestHeaderMapImpl{{"Addr", "x-foo-key;x-foo-key"}})
+                ->name(),
+            "foo_routes");
+}
```
Comment (AndresGuedez) [inline test/common/router/scoped_rds_test.cc]: I suggest also testing the edge case where the scope key is reused but the `route_configuration_name` is new.

File: source/common/router/scoped_rds.h [hunk 3]
```diff
@@ -133,8 +133,9 @@ class ScopedRdsConfigSubscription : public Envoy::Config::DeltaConfigSubscriptio
                          std::vector<std::string>& exception_msgs);
   // Removes given scopes from the managed set of scopes.
   // Returns true if any scope updated, false otherwise.
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply source/common/router/scoped_rds.h]: > Please update comment.

oopps, done.

File: source/common/router/scoped_rds.cc [hunk 4]
```diff
@@ -184,12 +184,15 @@ bool ScopedRdsConfigSubscription::addOrUpdateScopes(
   return any_applied;
 }
 
-bool ScopedRdsConfigSubscription::removeScopes(
+std::list<std::unique_ptr<ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper>>
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply source/common/router/scoped_rds.cc]: it's only used within this module, I would leave it if you don't feel strongly.

File: test/common/router/scoped_rds_test.cc (hunk 1 above)
Comment (stevenzzzz) [reply test/common/router/scoped_rds_test.cc]: done

File: test/common/router/scoped_rds_test.cc [hunk 5]
```diff
@@ -414,7 +414,7 @@ route_configuration_name: foo_routes
             "foo_routes");
 }
 
-// Tests that conflict resources are detected.
+// Tests that conflict resources in the same push are detected.
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply test/common/router/scoped_rds_test.cc]: done

File: test/common/router/scoped_rds_test.cc (hunk 2 above)
Comment (stevenzzzz) [reply test/common/router/scoped_rds_test.cc]: added a scenario where new scope has the same key but different route-table name that  will (conflict/remove-old-add-new) with existing scope.

**Pull request merged**